core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -756,19 +796,116 @@ public CompletableFuture<Optional<Throwable>> releaseAcquiredRecords(String memb
+                        // yet to be acknowledged, and thus should not be removed. Only a subMap will be removed from the
+                        // cachedState. The logic to remove batches from cachedState is as follows -
+                        //    Only full batches can be removed from the cachedState, For example if there is batch (0-99)
+                        //    and 0-49 records are acknowledged (ACCEPT or REJECT), the first 50 records will not be removed
+                        //    from the cachedState. Instead, the startOffset will be moved to 50, but the batch will only
+                        //    be removed once all the messages (0-99) are acknowledged (ACCEPT or REJECT)
+
+                        // Since only a subMap will be removed, we need to find the first and last keys of that subMap
+                        long firstKeyToRemove = cachedState.firstKey();
+                        long lastKeyToRemove;
+                        NavigableMap.Entry<Long, InFlightBatch> entry = cachedState.floorEntry(lastOffsetAcknowledged);
+                        if (lastOffsetAcknowledged == entry.getValue().lastOffset) {
+                            startOffset = cachedState.higherKey(lastOffsetAcknowledged);
+                            lastKeyToRemove = entry.getKey();
+                        } else {
+                            startOffset = lastOffsetAcknowledged + 1;
+                            if (entry.getKey().equals(cachedState.firstKey())) {
+                                // If the first batch in cachedState has some records yet to be acknowledged,
+                                // then nothing should be removed from cachedState
+                                lastKeyToRemove = -1;
+                            } else {
+                                lastKeyToRemove = cachedState.lowerKey(entry.getKey());
+                            }
+                        }
+
+                        if (lastKeyToRemove != -1) {
+                            NavigableMap<Long, InFlightBatch> subMap = cachedState.subMap(firstKeyToRemove, true, lastKeyToRemove, true);
+                            for (Long key : subMap.keySet()) {
+                                cachedState.remove(key);
+                            }
+                        }
+                    }
+                }
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
     private AcquiredRecords acquireNewBatchRecords(String memberId, long baseOffset, long lastOffset, long nextOffset) {
         lock.writeLock().lock();
         try {
             // Schedule acquisition lock timeout for the batch.
             TimerTask timerTask = scheduleAcquisitionLockTimeout(memberId, baseOffset, lastOffset);
             // Add the new batch to the in-flight records along with the acquisition lock timeout task for the batch.
+            boolean isCachedStateEmpty = cachedState.isEmpty();
```

Comment:
nit: we might not want to have a variable for just one condition and with one usage.